sdk/nodejs/runtime/closure/codePaths.ts

```diff
@@ -179,78 +193,118 @@ function searchUp(currentDir: string, fileToFind: string): string | null {
     return searchUp(parentDir, fileToFind);
 }
 
+/**
+ * findWorkspaceRoot detects if we are in a yarn/npm workspace setup, and
+ * returns the root of the workspace. If we are not in a workspace setup, it
+ * returns null.
+ */
+async function findWorkspaceRoot(programDirectory: string): Promise<string | null> {
+    let currentDir = upath.dirname(programDirectory);
+    let nextDir = upath.dirname(currentDir);
+    while (currentDir !== nextDir) {
+        const p = upath.join(currentDir, "package.json");
+        if (!fs.existsSync(p)) {
+            currentDir = nextDir;
+            nextDir = upath.dirname(currentDir);
+            continue;
+        }
+        const workspaces = parseWorkspaces(p);
+        for (const workspace of workspaces) {
+            const files = await pGlob(upath.join(currentDir, workspace));
+            const normalized = upath.normalizeTrim(programDirectory);
+            if (files.map((f) => upath.normalizeTrim(f)).includes(normalized)) {
+                return currentDir;
+            }
+        }
+        return null;
+    }
+    return null;
+}
+
+function parseWorkspaces(packageJsonPath: string): string[] {
+    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf8"));
+    if (packageJson.workspaces && Array.isArray(packageJson.workspaces)) {
+        return packageJson.workspaces;
+    }
+    if (packageJson.workspaces?.packages && Array.isArray(packageJson.workspaces.packages)) {
+        return packageJson.workspaces.packages;
+    }
+    return [];
+}
+
 // allFolders computes the set of package folders that are transitively required by the root
 // 'dependencies' node in the client's project.json file.
-function allFoldersForPackages(
+async function allFoldersForPackages(
     includedPackages: Set<string>,
     excludedPackages: Set<string>,
     logResource: Resource | undefined,
 ): Promise<Set<string>> {
-    return new Promise((resolve, reject) => {
-        // the working directory is the directory containing the package.json file
-        const workingDir = searchUp(".", "package.json");
-        if (workingDir === null) {
-            // we couldn't find a directory containing package.json
-            // searching up from the current directory
-            throw new ResourceError("Failed to find package.json.", logResource);
+    // the working directory is the directory containing the package.json file
+    let workingDir = searchUp(".", "package.json");
+    if (workingDir === null) {
+        // we couldn't find a directory containing package.json
+        // searching up from the current directory
+        throw new ResourceError("Failed to find package.json.", logResource);
+    }
+    // Ensure workingDir is a relative path so we get relative paths in the output.
+    workingDir = upath.relative(upath.resolve("."), workingDir);
```

Comment:
This was not the case previously, and the uploaded lambda ended up dropping the dependencies.